File: app/src/main/java/algorithm/study/dohyun/boj/Boj_12789.java
package algorithm.study.dohyun.boj;
import java.util.*;
import java.io.*;
import java.util.Queue;
import java.util.Stack;

public class Boj_12789 {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        Queue<Integer> q = new LinkedList<>();
        Stack<Integer> stk = new Stack<>();
        StringTokenizer st = new StringTokenizer(br.readLine());
        for(int i= 0; i< N; i++)
            q.add(Integer.parseInt(st.nextToken()));
        int num = 1;
        while( !q.isEmpty() ) {
            if (q.peek().equals(num)) {
                q.poll();
                num++;
            } else if (!stk.isEmpty() && stk.peek().equals(num)) {
                stk.pop();
                num++;
            } else {
                stk.push(q.poll());
            }
        }
        while( ! stk.isEmpty() ){
            if( stk.peek().equals(num) ) {
                stk.pop();
                num++;
            }
            else{
                System.out.println("Sad");
                return;
            }
        }
        System.out.println("Nice");

    }
}

File: app/src/main/java/algorithm/study/devjk_sample/baekjoon/Boj_12789.kt
package algorithm.study.devjk_sample.baekjoon

import java.util.*

/**
 * 백준 - 도키도키 간식드리미
 *
 * @link https://www.acmicpc.net/problem/12789
 */
class Boj_12789 {

    fun solve() {
        val n = readln().toInt()
        val arr = readln().split(" ").map { num -> num.toInt() }
        val q: Queue<Int> = LinkedList(arr)
        val st = Stack<Int>()
        var target = 1
        while (!q.isEmpty()) {
            if (!st.empty() && st.peek() == target) {
                st.pop()
                target++
                continue
            }
            val cur = q.poll()
            if (cur == target) {
                target++
                continue
            }
            st.push(cur)
        }
        while (!st.empty()) {
            val cur = st.pop()
            if (cur != target) {
                print("Sad")
                return
            }
            target++
        }
        print("Nice")
    }
}

fun main(args: Array<String>) {
    val sol = Boj_12789()
    sol.solve()
}